confdgnmi/src/confd_gnmi_demo_adapter.py

```diff
@@ -39,51 +85,218 @@ def _fill_demo_db(self):
-                                          atomic=False)
-
-            response = gnmi_pb2.SubscribeResponse(update=notif)
-            log.debug("<== response=%s", response)
-            return response
-
-    def get_subscription_handler(self) -> SubscriptionHandler:
+
+            self.adapter.db_lock.acquire()
+            get_updates(self.adapter.demo_db)
+            # 'if' below is optimization
+            if len(update) == 0:
+                get_updates(self.adapter.demo_state_db)
+            self.adapter.db_lock.release()
+
+            log.debug("<== update=%s", update)
+            return update
+
+        def get_monitored_changes(self) -> []:
+            self.change_db_lock.acquire()
+            log.debug("==> self.change_db=%s", self.change_db)
+            assert len(self.change_db) > 0
+            update = []
+            for c in self.change_db:
+                prefix_str = make_xpath_path(
+                    gnmi_prefix=self.subscription_list.prefix)
+                p = c[0]
+                if c[0].startswith(prefix_str):
+                    p = c[0][len(prefix_str):]
+                update.append(gnmi_pb2.Update(path=make_gnmi_path(p),
+                                              val=gnmi_pb2.TypedValue(
+                                                  string_val=c[1])))
+            self.change_db = []
+            self.change_db_lock.release()
+            log.debug("<== update=%s", update)
+            return update
+
+        def _get_random_changes(self):
+            log.debug("==>")
+            changes = []
+            candidate_paths = set()
+            for mp in self.monitored_paths:
+                for p, v in self.adapter.demo_db.items():
+                    if p.startswith(mp):
+                        # we only simulate changes on type leaf
+                        candidate_paths.add(
+                            p.replace("/name", "/type"))
+            log.debug("candidate_paths=%s", candidate_paths)
+            for path in random.sample(candidate_paths,
+                                      min(len(candidate_paths), 4)):
+                new_val = "gigabitEthernet"
+                if self.adapter.demo_db[path] == "gigabitEthernet":
+                    new_val = "fastEther"
+                log.debug("adding change path=%s, new_val=%s", path,
+                          new_val)
+                changes.append((path, new_val))
+            if randint(0, 9) < 8:
+                changes.append("send")
+            else:
+                changes.append("break")
+            log.debug("<== changes=%s", changes)
+            return changes
+
+        @staticmethod
+        def _get_config_changes():
+            log.debug("==>")
+            assert "changes_idx" in GnmiDemoServerAdapter.config
+            assert "changes" in GnmiDemoServerAdapter.config
+            changes = []
+            idx = GnmiDemoServerAdapter.config["changes_idx"]
```

Comment:
repetetive access into `len(UglyLongNameAdapter.config["changes"])` breaks DRY principle.
make local variable, use further...

Reply:
Yes. I'll keep it as it is only 2 cases.